### ISSUE TITLE
before_destroy callback is not called with has_many :through association.

### ISSUE THREAD

**Issue by dbwinger** [sferik/rails_admin#2281]
It seems RailsAdmin is somehow preventing the `before_destroy` callback from being run on the association.

Here are my models (relevant code):
```ruby
class Race < ActiveRecord::Base
  ...
  has_many :race_distances, dependent: :destroy
  has_many :distances, -> { order('reviews_count DESC') }, through: :race_distances
  ...
end

class RaceDistance < ActiveRecord::Base
  belongs_to :distance
  belongs_to :race

  before_destroy { raise Exception.new 'before destroy' }
  ...
end

class Distance < ActiveRecord::Base
  has_many :races, through: :race_distances
  has_many :race_distances
end
```
Nothing custom in rails admin config.  So, editing a Race record, I can add/remove "distances" (RaceDistance) records fine using the default filterable multiselect.  BUT, the before_destroy callback never gets called.  I also tried overriding `RaceDistance#destroy`, `RaceDistance#delete`, `RaceDistance.destroy_all` and `RaceDistance.delete_all` and it appears none of them are getting called.  So is RailsAdmin doing something at a lower level than AR in this case?  I confirmed for another `has_many` association it does get called.

Any thoughts on why, or where I should look in the rails_admin code?  I tried to look around but was a bit lost to where the deletion of nested associations is happening.

**Comment by mshibuya** [comment]
RA does nothing special about it, you'd better check ActiveRecord association implementation.

**Comment by coreyaus** [comment]
I experienced this same issue and explain how I overcame it in point 1 below. I've pasted the relevant code from my models at the bottom.

It does seem like there is something odd happening as the process was working through the front_end. Whenever project creators removed a project collaborator through the frontend process (which uses simple_form) the project state was updated by the `after_commit, on: :destroy` callback. Also, destroying a collaborator role through the console would trigger the callback successfully. However, when an admin removed a collaborator through RailsAdmin it didn't trigger the callback. This may not be a bug so I'm not re-opening this issue - just adding an update here for anyone who has a similar issue in the future.

1. In the end I achieved what I needed by adding an `after_remove` callback directly on the relevant association (in my case that was the `collaborators` association in the Project model). The `after_remove` callback is successfully triggered whenever a collaborator is removed from a Project through RailsAdmin.
2. I found the discussion of this open Rails issue interesting and relevant when reviewing this callbacks configuration to get RailsAdmin behaving as I needed: https://github.com/rails/rails/issues/14365. It outlines the opposite of this issue (where `after_remove` was not working as expected). The relevance is simply that it reveals some differences between how callbacks are handled if the object is destroyed directly (e.g. `associated_object.destroy`). or through its association (e.g. `object.collection.delete(associated_object)`)

I would be more than happy to hear any explanations people have about why my original code below didn't work (that is, why the `after_commit, on: :destroy` callback was not triggered when removing a collaborator through RailsAdmin). Perhaps my expectation was simply incorrect. In any event, hopefully anyone else with a similar issue has more luck with `after_remove` callbacks on the association.

```
class User < ApplicationRecord
  has_many :roles, dependent: :destroy
  has_many :projects, through: :roles, dependent: :destroy
  ...
end

class Project < ApplicationRecord
  has_many :roles, dependent: :destroy
  has_many :users, through: :roles
  has_many :creators, -> { merge(Role.creators) }, through: :roles
  has_many :collaborators, -> { merge(Role.collaborators) }, through: :roles
  ...
end

class Role < ApplicationRecord
  belongs_to :user
  belongs_to :project
  scope :creators, -> { where(role: 'creator') }
  scope :collaborators, -> { where(role: 'collaborator') }
  after_commit :update_project_state, on: :destroy, if: -> (r) { r.role == 'collaborator' }
  ...
  private
    def update_project_state
    end
    ...
end
```
